src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1151,7 +1151,17 @@ private SolutionCompilationState UpdateAnalyzerConfigDocumentState(StateChange s
     private bool TryGetCompilationTracker(ProjectId projectId, [NotNullWhen(returnValue: true)] out ICompilationTracker? tracker)
         => _projectIdToTrackerMap.TryGetValue(projectId, out tracker);
 
-    private static readonly Func<ProjectId, SolutionState, RegularCompilationTracker> s_createCompilationTrackerFunction = CreateCompilationTracker;
+    private static readonly Func<ProjectId, (SolutionState, bool), RegularCompilationTracker> s_createCompilationTrackerFunction = CreateCompilationTracker;
+
+    private static RegularCompilationTracker CreateCompilationTracker(ProjectId projectId, (SolutionState solution, bool requiredSourceGeneratorDocumentsOnly) data)
+    {
+        var tracker = CreateCompilationTracker(projectId, data.solution);
+        if (data.requiredSourceGeneratorDocumentsOnly)
+        {
+            tracker = tracker.WithDoNotCreateCreationPolicy();
+        }
```

Comment:
1. doc what's going on here.
2. instead of 'With'in the result, i think i'd prefer that CreateCompilationTracker takes in a value controlling this.